Add numeric_debugger top level APIs

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130643

Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Add `prepare_for_propagation_comparison` and `extract_results_from_loggers`

Summary:
Add two top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dulinriley commented on test/quantization/pt2e/test_numeric_debugger.py:
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
+        res = m_logger(*example_inputs)
+
+        from torch.ao.quantization.pt2e.numeric_debugger import OutputLogger
+        assert any(isinstance(m, OutputLogger) for m in m_logger.modules())
```
`any` is not quite specific, perhaps we should check `all` instead? Or maybe find a node of a particular function, and make sure that a call_module after it is an OutputLogger?

jerryzh168 replied:
sure, we can't check all because not all modules are output loggers, but I can check the node name of loggers contains conv2d

dulinriley commented on test/quantization/pt2e/test_numeric_debugger.py:
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
+        from torch.ao.quantization.pt2e.numeric_debugger import OutputLogger
+        assert any(isinstance(m, OutputLogger) for m in m_logger.modules())
+
+        self.assertEqual(res, ref)
```
You should use torch.testing.assert_close instead

jerryzh168 replied:
why? adding logger is not going to change the result right

dulinriley commented on test/quantization/pt2e/test_numeric_debugger.py:
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
+        quantizer = XNNPACKQuantizer().set_global(
+            get_symmetric_quantization_config(is_per_channel=False)
+        )
+        m = prepare_pt2e(m, quantizer)
```
You might want to have some kind of check/assert in prepare_pt2e to ensure the loggers *aren't* in there.
The loggers will block both dead code elimination, pattern matching, and optimizations that remove nodes

jerryzh168 replied:
prepare_pt2e should not be aware of `OutputLogger` I think, this has to be enforced by user I feel

dulinriley commented on test/quantization/pt2e/test_numeric_debugger.py:
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
+        m = prepare_pt2e(m, quantizer)
+        m(*example_inputs)
+        m = convert_pt2e(m)
+        debug_handle_map = _extract_debug_handles(m)
```
Why is this API private (leading underscore)?

jerryzh168 replied:
this is a test-only helper function, not user facing API

jerryzh168 replied:
sorry, this can be removed now actually

dulinriley commented on test/quantization/pt2e/test_numeric_debugger.py:
```diff
@@ -121,3 +123,40 @@ def test_re_export_preserve_handle(self):
+        results = extract_results_from_loggers(m_ref_logger, m_quant_logger)
+        for node, node_summary in results.items():
+            if len(node_summary.results) > 0:
+                self.assertTrue(node_summary.results[0].sqnr >= 35)
```
```suggestion
                self.assertGreaterEqual(node_summary.results[0].sqnr, 35)
```

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+
+
+class OutputLogger(torch.nn.Module):
+    """
+    Base class for capturing output values.
+    """
```
Can you add a better description? What types of output values does it capture? When would it not work?
I'd recommend focusing on the cases specific to prepare/convert for now

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
             unique_id += 1
+
+
+class OutputLogger(torch.nn.Module):
```
Does this need a `backwards_compatibility` annotation? Is it a public torch API?

jerryzh168 replied:
this is a public API, I'll add it to `torch.ao.quantization`, but we have not been using `backwards_compatibility` decorators.

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+    def __init__(
+        self,
+        node_name: str,
+        nn_module_stack: object,
```
You could make this default to None, in case other users don't have a stack available. It's an optional part of the logger, just to make identifying stuff easier. Same with `node_name` too, only `debug_handle` is required

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+            self.stats.append(x.detach())
+        return x
+
+    def __repr__(self) -> str:
+        clean_dict = {
+            k: v
+            for k, v in self.__dict__.items()
+            # skip torch.nn.Module keys
+            if (k != "training") and not k.startswith("_")
+        }
+        return f"OutputLogger({clean_dict})"
```
Why include all this stuff?
Also, I think typically subclasses of Module use "extra_repr" instead
```suggestion
    def __repr__(self) -> str:
        return f"OutputLogger(node_name={self.node_name}, nn_module_stack={self.nn_module_stack}, debug_handle={self.debug_handle}, num_stats={len(self.stats)})"
```

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+        return f"OutputLogger({clean_dict})"
+
+
+def _insert_logger(model: GraphModule, node: Node, debug_handle: str) -> None:
```
nit: might want to return the OutputLogger instance, or the logger node. Some callers might want to access it after-the-fact

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+    # add a logger after the node
+    with model.graph.inserting_after(node):
+        get_new_attr_name = get_new_attr_name_with_prefix(f"{node.name}_logger")
+        logger_name = get_new_attr_name(model)
```
Maybe assert that this attribute name doesn't exist yet?

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+class QuantizationComparisonResult:
+    actual: torch.Tensor
+    ref: torch.Tensor
+
```
I recommend adding `__post_init__` and assert that the actual/ref are really tensors

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+@dataclass(frozen=True)
+class NodeAccuracySummary:
+    handle: str
+    actual_node_name: str
+    actual_module_stack: str
+    ref_node_name: str
+    ref_module_stack: str
```
nit: swap this order.

Also, for the expansion of this concept to "one ref node, possibly multiple comparison nodes", you can have a list of actual_name and actual_stack.

But that can be a follow-up PR if you want
```suggestion
    ref_node_name: str
    ref_module_stack: str
    actual_node_name: str
    actual_module_stack: str
```

jerryzh168 replied:
I feel we can design this part when there is a real use case

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+        return s
+
+
+def _module_stack_to_str(module_stack: object) -> str:
```
You might want to put this function somewhere else, I'm sure other people like to visualize the nn_module_stack

jerryzh168 replied:
this seems to be a small util, we can move when other people also needs this I think

dulinriley commented on torch/ao/quantization/pt2e/numeric_debugger.py:
```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
+    ref_handles: Dict[str, Tuple[str, object, List[torch.Tensor]]] = {}
+    actual_handles: Dict[str, Tuple[str, object, List[torch.Tensor]]] = {}
+    for _name, module in ref_model.named_children():
+        if isinstance(module, OutputLogger):
```
I recommend checking if `len(module.stats) > 0` here, to avoid comparing empty loggers. Maybe a warning can be emitted in that case, the user didn't run `forward`

jerryzh168 committed:
Update on "Add `prepare_for_propagation_comparison` and `extract_results_from_loggers`"


Summary:
Add two top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add `prepare_for_propagation_comparison` and `extract_results_from_loggers`"


Summary:
Add two top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add `prepare_for_propagation_comparison` and `extract_results_from_loggers`"


Summary:
Add two top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add `prepare_for_propagation_comparison` and `extract_results_from_loggers`"


Summary:
Add two top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add `prepare_for_propagation_comparison` and `extract_results_from_loggers`"


Summary:
Add two top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add numeric_debugger top level APIs"


Summary:
Add three top level APIs for numeric debugger in pt2e flow that can log intermediate output in the model
and calculate summary for metric comparisons between nodes in two graphs

* `prepare_for_propagation_comparison`
* `extract_results_from_loggers`
* `compare_results`

Test Plan:
python test/test_quantization.py -k test_prepare_for_propagation_comparison
python test/test_quantization.py -k test_extract_results_from_loggers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge -f "low risk, one job is waiting in queue and timed out"
